### PR TITLE
Added cost function parameter to role_assignment

### PR DESCRIPTION
**Pull request by mmurley3**
addresses #923

### CODE REVIEW

**Comment by mmurley3** [comment]
Not sure if this overlaps with #1019 and/or #1018

**Comment by JNeiger** [comment]
I think having both would be a good idea. #1018 deals with how many robots a play should be assigned while this cost function deals with which specific robot is assigned to that role.

On that note, we may want to pass in the specific robot the cost is being calculated for into the lambda function

**Comment by mmurley3** [comment]
Not sure if this is exactly what you were talking about Joe, but give it a glance. A bit late, I know.

**Comment by JNeiger** [comment]
That's actually pretty close considering how vague I was. 

I was thinking that for the cost function itself, having the specific robot the cost is being evaluated for, may be useful.

Eg:
```
if req.cost_func != none:
     cost += req.cost_func(robot)
```

One example of a use case for this is that when we want to find a receiver for a pass, we may want to try and get a receiver on our side of the ball so it is easier for the them pass or shoot directly afterwards. If we pass in the robot as an argument, it becomes a pretty simple problem to solve.

On top of that, the cost function will still continue to have the same functionality as described in the original issue since you can just ignore that argument.

Thoughts?

**Comment by mmurley3** [comment]
Ah, I see what you mean.  I'll get it up in a bit.

**Comment by mmurley3** [comment]
Ok, it should be good now.  I tested a few cases and it should work as long as people declare the robot parameter.  Let me know if there's anything that I missed.

File: soccer/gameplay/role_assignment.py [hunk 1]
```diff
@@ -254,6 +265,8 @@ def fail(errStr):
                     cost += req.robot_change_cost
                 if not robot.has_chipper():
                     cost += req.chipper_preference_weight
+                if req.cost_func != None:
```
Comment (jgkamat) [inline soccer/gameplay/role_assignment.py]: `const_func is not None`

File: soccer/gameplay/role_assignment.py [hunk 2]
```diff
@@ -21,6 +21,10 @@ def __init__(self):
 
         # multiply this by the distance between two points to get the cost
         self.position_cost_multiplier = 1.0
+        
+        # A lambda function property that allows customization of cost
+        # Has exactly one parameter, which is a robot
+        self.cost_func = None
```
Comment (jcarn) [inline soccer/gameplay/role_assignment.py]: What if we did self.cost_func = lambda r: 0

File: soccer/gameplay/role_assignment.py (hunk 1 above)
Comment (jcarn) [inline soccer/gameplay/role_assignment.py]: Then we don't need this check

File: soccer/gameplay/role_assignment.py (hunk 2 above)
Comment (jcarn) [reply soccer/gameplay/role_assignment.py]: This also gives a bit of an idea about what a function should look like, good for readability

**Pull request merged**